src/AzureExtension/DataModel/DataObjects/Project.cs

```diff
@@ -1,186 +1,232 @@
+    {
+        return new Project
+        {
+            InternalId = project.Id.ToString(),
+            Name = project.Name ?? string.Empty,
+            Description = project.Description ?? string.Empty,
+            OrganizationId = organizationId,
+            TimeUpdated = DateTime.Now.ToDataStoreInteger(),
+        };
+    }
+
+    private static Project CreateFromTeamProject(TeamProjectReference project, long organizationId)
+    {
+        return new Project
+        {
+            InternalId = project.Id.ToString(),
+            Name = project.Name ?? string.Empty,
+            Description = project.Description ?? string.Empty,
+            OrganizationId = organizationId,
+            TimeUpdated = DateTime.Now.ToDataStoreInteger(),
+        };
+    }
+
+    public static Project AddOrUpdateProject(DataStore dataStore, Project project)
+    {
+        // Check for existing Project data.
+        var existingProject = GetByInternalId(dataStore, project.InternalId);
+        if (existingProject is not null)
+        {
+            // Many of the same Project records will be created on a sync, and to
+            // avoid unnecessary updating and database operations for data that
+            // is extremely unlikely to have changed in any significant way, we
+            // will only update every UpdateThreshold amount of time.
+            if ((project.TimeUpdated - existingProject.TimeUpdated) > _updateThreshold)
+            {
+                project.Id = existingProject.Id;
+                dataStore.Connection!.Update(project);
+                project.DataStore = dataStore;
+                return project;
+            }
+            else
+            {
+                return existingProject;
+            }
+        }
+
+        // No existing project, add it.
+        project.Id = dataStore.Connection!.Insert(project);
+        project.DataStore = dataStore;
+        return project;
+    }
+
+    // Direct Get always returns a non-null object for reference in other objects.
+    public static Project Get(DataStore? dataStore, long id)
+    {
+        if (dataStore == null)
+        {
+            return new Project();
+        }
+
+        var project = dataStore.Connection!.Get<Project>(id);
+        if (project != null)
+        {
+            project.DataStore = dataStore;
+        }
+
+        return project ?? new Project();
+    }
+
+    public static Project? Get(DataStore? dataStore, string projectName, string organizationName)
+    {
+        if (dataStore == null)
+        {
+            return null;
+        }
+
+        // This is not guaranteed to only return one result, such as the case of multiple
+        // organization records with different connections.
+        var sql = @"SELECT * FROM Project AS P WHERE (P.Name = @Name OR P.InternalId = @Name) AND P.OrganizationId IN (SELECT Id FROM Organization WHERE Organization.Name = @Org)";

```

Comment:
NIT: Declare the columns instead of using *.

Reply:
I do appreciate the sentiment and normally agree with that, but I tried this before and it encountered some odd problems and bugs with the ORM and made even small changes a big PITA and likely to break the data model and extension at runtime. This is an intentional choice to avoid nefarious bugs if a new property is added, removed or renamed from a data object and keep the sql statements relatively simple and let the ORM do its thing. The IDE will not spot errors in naming of columns with what is in the SQL strings. I think it actually improves both maintainability and readability of the code to keep it with *, as the ORM will do the right thing and create the data object based on the fields presented.  The data model has been very stable and I'd prefer to not risk it by changing this without a very good functional reason. All the other data objects function the same way without issue.